### ISSUE TITLE
Dependency graph for VMSS prevents Loadbalancer update

### ISSUE THREAD

**Issue by mbrancato** [terraform-providers/terraform-provider-azurerm#4769]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.12.3
+ provider.azurerm v1.36.1
```

### Affected Resource(s)

* `azurerm_virtual_machine_scale_set`
* `azurerm_lb_probe`
* `azurerm_lb`

The VMSS is dependent on the LB probe and LB resources. When certain changes are made to these causing recreation, the `apply` will fail. To correct this, I see two possible solutions:

1. Modify the VMSS to remove the LB probe dependency, then re-add once the LB probe is recreated.
2. Delete the VMSS first, then recreate. Obviously this could have a larger impact for a small change.

### Terraform Configuration Files

```hcl
variable "resource_group" {}

data "azurerm_resource_group" "test" {
  name = "${var.resource_group}"
}

locals {
  prefix = "test-vmss-bug"
}

resource "azurerm_virtual_network" "test" {
  name                = "${local.prefix}-vnet"
  address_space       = ["10.0.0.0/16"]
  location            = "${data.azurerm_resource_group.test.location}"
  resource_group_name = "${data.azurerm_resource_group.test.name}"
}

resource "azurerm_subnet" "test" {
  name                 = "${local.prefix}-subnet"
  resource_group_name  = "${data.azurerm_resource_group.test.name}"
  virtual_network_name = "${azurerm_virtual_network.test.name}"
  address_prefix       = "10.0.2.0/24"
}

resource "azurerm_lb" "test" {
  name                = "${local.prefix}-lb"
  location            = "${data.azurerm_resource_group.test.location}"
  resource_group_name = "${data.azurerm_resource_group.test.name}"
  #sku                 = "Standard"

  frontend_ip_configuration {
    name                          = "${local.prefix}-lbip"
    subnet_id                     = "${azurerm_subnet.test.id}"
    private_ip_address            = "10.0.2.250"
    private_ip_address_allocation = "Static"
  }
}

resource "azurerm_lb_backend_address_pool" "test" {
  resource_group_name = "${data.azurerm_resource_group.test.name}"
  loadbalancer_id     = "${azurerm_lb.test.id}"
  name                = "BackEndAddressPool"
}

resource "azurerm_lb_nat_pool" "lbnatpool" {
  resource_group_name            = "${data.azurerm_resource_group.test.name}"
  name                           = "ssh"
  loadbalancer_id                = "${azurerm_lb.test.id}"
  protocol                       = "Tcp"
  frontend_port_start            = 50000
  frontend_port_end              = 50119
  backend_port                   = 22
  frontend_ip_configuration_name = "${local.prefix}-lbip"
}

resource "azurerm_lb_probe" "test" {
  resource_group_name = "${data.azurerm_resource_group.test.name}"
  loadbalancer_id     = "${azurerm_lb.test.id}"
  name                = "http-probe"
  protocol            = "Http"
  request_path        = "/health"
  port                = 8080
}

resource "azurerm_lb_rule" "azlb" {
  resource_group_name            = "${data.azurerm_resource_group.test.name}"
  loadbalancer_id                = "${azurerm_lb.test.id}"
  name                           = "${local.prefix}-lbrule"
  protocol                       = "Tcp"
  frontend_port                  = "${azurerm_lb_probe.test.port}"
  backend_port                   = "${azurerm_lb_probe.test.port}"
  frontend_ip_configuration_name = "${local.prefix}-lbip"
  enable_floating_ip             = false
  backend_address_pool_id        = "${azurerm_lb_backend_address_pool.test.id}"
  idle_timeout_in_minutes        = 5
  probe_id                       = "${azurerm_lb_probe.test.id}"
}


resource "azurerm_virtual_machine_scale_set" "test" {
  name                 = "${local.prefix}-vmss"
  location             = "${data.azurerm_resource_group.test.location}"
  resource_group_name  = "${data.azurerm_resource_group.test.name}"
  automatic_os_upgrade = true
  upgrade_policy_mode  = "Rolling"

  rolling_upgrade_policy {
    max_batch_instance_percent              = 20
    max_unhealthy_instance_percent          = 20
    max_unhealthy_upgraded_instance_percent = 5
    pause_time_between_batches              = "PT0S"
  }

  health_probe_id = "${azurerm_lb_probe.test.id}"

  sku {
    name     = "Standard_A2"
    tier     = "Standard"
    capacity = 2
  }

  storage_profile_image_reference {
    publisher = "Canonical"
    offer     = "UbuntuServer"
    sku       = "18.04-LTS"
    version   = "latest"
  }

  storage_profile_os_disk {
    name              = ""
    caching           = "ReadWrite"
    create_option     = "FromImage"
    managed_disk_type = "Standard_LRS"
  }

  os_profile {
    computer_name_prefix = "${local.prefix}-vm"
    admin_username       = "legituser"
    admin_password       = "TestPassword123!!"
  }

  network_profile {
    name    = "terraformnetworkprofile"
    primary = true

    ip_configuration {
      name                                   = "IPConfiguration"
      primary                                = true
      subnet_id                              = "${azurerm_subnet.test.id}"
      load_balancer_backend_address_pool_ids = ["${azurerm_lb_backend_address_pool.test.id}"]
      load_balancer_inbound_nat_rules_ids    = ["${azurerm_lb_nat_pool.lbnatpool.id}"]
    }
  }
}
```

### Expected Behavior

Terraform should be able to find a path to update the dependencies of the VMSS.

### Actual Behavior

```
azurerm_lb_nat_pool.lbnatpool: Destroying... [id=/subscriptions/11112222-3333-
4444-5555-666677778888/resourceGroups/example-rg/providers/Microsoft.Network
/loadBalancers/test-vmss-bug-lb/inboundNatPools/ssh]
azurerm_lb_rule.azlb: Destroying... [id=/subscriptions/11112222-3333-4444-5555-
666677778888/resourceGroups/example-rg/providers/Microsoft.Network
/loadBalancers/test-vmss-bug-lb/loadBalancingRules/test-vmss-bug-lbrule]

Error: Error creating/updating Load Balancer "test-vmss-bug-lb" (Resource Group 
"example-rg"): network.LoadBalancersClient#CreateOrUpdate: Failure sending 
request: StatusCode=400 -- Original Error: 
Code="LoadBalancerInboundNatPoolInUseByVirtualMachineScaleSet" Message="Cannot 
remove inbound nat pool ssh from load balancer since it is in use by virtual 
machine scale set /subscriptions/11112222-3333-4444-5555-666677778888
/resourceGroups/example-rg/providers/Microsoft.Compute/virtualMachineScaleSets
/test-vmss-bug-vmss." Details=[]



Error: Error Creating/Updating Load Balancer "test-vmss-bug-lb" (Resource Group 
"example-rg"): network.LoadBalancersClient#CreateOrUpdate: Failure sending 
request: StatusCode=400 -- Original Error: 
Code="CannotRemoveRuleUsedByProbeUsedByVMSS" Message="Load balancer rule 
/subscriptions/11112222-3333-4444-5555-666677778888/resourceGroups/example-rg
/providers/Microsoft.Network/loadBalancers/test-vmss-bug-lb/loadBalancingRules
/test-vmss-bug-lbrule cannot be removed because the rule references the load 
balancer probe /subscriptions/11112222-3333-4444-5555-666677778888
/resourceGroups/example-rg/providers/Microsoft.Network/loadBalancers
/test-vmss-bug-lb/probes/http-probe that is used as health probe by VM scale 
set /subscriptions/11112222-3333-4444-5555-666677778888/resourceGroups
/example-rg/providers/Microsoft.Compute/virtualMachineScaleSets
/test-vmss-bug-vmss. To remove this rule, please update VM scale set to remove 
the reference to the probe." Details=[]
```

### Steps to Reproduce

1. `terraform apply`
2. Change something forcing the recreate of the LB or LB probe resources. Example provided is changing the LB `sku`.
3. `terraform apply`

### Important Factoids

The probe is required for Rolling updates. If the probe is to be removed from the VMSS, the update method probably needs to be temporarily changed for this to work.

### References

* #4398

**Comment by tombuildsstuff** [comment]
hi @mbrancato 

Thanks for opening this issue.

In the Terraform Configuration above there's no (direct) dependency between the `azurerm_lb_rule` resource and the `azurerm_virtual_machine_scale_set` resource - as such Terraform believes that it can modify these at the same time. As such you should be able to add this using a `depends_on` attribute to ensure that the VM Scale Set is modified before the Load Balancer Rule.

It's worth nothing that the existing `azurerm_virtual_machine_scale_set` resource is being superseded in 2.0 (as detailed in #2807) which modifies how updates happen so that we can roll the instances/profile of the scale set if required. However this too requires defining a `depends_on` from the `azurerm_linux_virtual_machine_scale_set` resource to the `azurerm_lb_rule`, due to the nature of how this is implemented at Azure's end (that there's no direct dependency from the VMSS to the LB Rule).

Given this is caused by a missing dependency you should be able to resolve this using a `depends_on` ([here's an example of how we do that in the tests](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_virtual_machine_scale_set_test.go#L5719)) - however since this is a question regarding Terraform Configuration rather than a bug in the Azure Provider I'm going to close this issue for the moment; but please let us know if that doesn't work for you and we'll take another look.

Thanks!

**Comment by mbrancato** [comment]
Hi @tombuildsstuff, adding `depends_on = [azurerm_lb_rule.azlb]` to the `azurerm_virtual_machine_scale_set` does not solve the problem here. I tried this before opening the case which is why I referenced it to the previous issue.

I've added the following block to my example given above and upgraded TF to 0.12.12, the same errors are shown and it doesn't change the plan.

```hcl
  depends_on = [
    azurerm_lb_rule.azlb,
    azurerm_lb.test,
    azurerm_lb_backend_address_pool.test,
    azurerm_lb_probe.test
  ]
```

**Comment by mbrancato** [comment]
@tombuildsstuff I've submitted a PR to add an acceptance test here. Running the test fails:

```hcl
$ go test '-run=^TestAccAzureRMVirtualMachineScaleSet_updateHealthProbe' -v
=== RUN   TestAccAzureRMVirtualMachineScaleSet_updateHealthProbe
=== PAUSE TestAccAzureRMVirtualMachineScaleSet_updateHealthProbe
=== CONT  TestAccAzureRMVirtualMachineScaleSet_updateHealthProbe
--- FAIL: TestAccAzureRMVirtualMachineScaleSet_updateHealthProbe (543.95s)
    testing.go:569: Step 1 error: errors during apply:
        
        Error: Error Creating/Updating Load Balancer "acctestlb-191101205623326072" (Resource Group "acctestRG-191101205623326072"): network.LoadBalancersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="CannotRemoveRuleUsedByProbeUsedByVMSS" Message="Load balancer rule /subscriptions/<subscription_id>/resourceGroups/acctestRG-191101205623326072/providers/Microsoft.Network/loadBalancers/acctestlb-191101205623326072/loadBalancingRules/AccTestLBRule cannot be removed because the rule references the load balancer probe /subscriptions/<subscription_id>/resourceGroups/acctestRG-191101205623326072/providers/Microsoft.Network/loadBalancers/acctestlb-191101205623326072/probes/acctest-lb-probe that is used as health probe by VM scale set /subscriptions/<subscription_id>/resourceGroups/acctestRG-191101205623326072/providers/Microsoft.Compute/virtualMachineScaleSets/acctvmss-191101205623326072. To remove this rule, please update VM scale set to remove the reference to the probe." Details=[]
        
        
FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-azurerm/azurerm	544.011s
```

**Comment by katbyte** [comment]
The tests in #5261 changing the load balancer now pass? so this might be fixed on the azure side?

**Comment by tombuildsstuff** [comment]
Since this test has been merged into #5261 (which switches this over to using the new VMSS Resource coming in 2.0) - which now passes, I'm going to close this issue for the moment and tag this with the 2.0 milestone - but ultimately the fix for this will become available in the coming weeks once the 2.0 Beta becomes available.
